// @navikt/ds-react/cjs/popover/Popover.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Popover = void 0;
const react_1 = require("@floating-ui/react");
const react_2 = __importStar(require("react"));
const Date_Input_1 = require("../date/Date.Input");
const Modal_context_1 = require("../modal/Modal.context");
const DismissableLayer_1 = require("../overlays/dismissablelayer/DismissableLayer");
const Theme_1 = require("../theme/Theme");
const hooks_1 = require("../util/hooks");
const useMergeRefs_1 = require("../util/hooks/useMergeRefs");
const PopoverContent_1 = __importDefault(require("./PopoverContent"));
/**
 * A component that displays a popover.
 *
 * @see [📝 Documentation](https://aksel.nav.no/komponenter/core/popover)
 * @see 🏷️ {@link PopoverProps}
 *
 * @example
 * ```jsx
 * <Button ref={buttonRef} onClick={() => setOpenState(true)}>
 *   Åpne popover
 * </Button>
 * <Popover
 *   open={openState}
 *   onClose={() => setOpenState(false)}
 *   anchorEl={buttonRef.current}
 * >
 *   <Popover.Content>Innhold her!</Popover.Content>
 * </Popover>
 * ```
 */
exports.Popover = (0, react_2.forwardRef)((_a, ref) => {
    var { className, children, anchorEl, arrow = true, open, onClose, placement = "top", offset, strategy: userStrategy, flip: _flip = true } = _a, rest = __rest(_a, ["className", "children", "anchorEl", "arrow", "open", "onClose", "placement", "offset", "strategy", "flip"]);
    const { cn } = (0, Theme_1.useRenameCSS)();
    const arrowRef = (0, react_2.useRef)(null);
    const isInModal = (0, Modal_context_1.useModalContext)(false) !== undefined;
    const datepickerContext = (0, Date_Input_1.useDateInputContext)(false);
    const chosenStrategy = userStrategy !== null && userStrategy !== void 0 ? userStrategy : (isInModal ? "fixed" : "absolute");
    const chosenFlip = datepickerContext ? false : _flip;
    const themeContext = (0, Theme_1.useThemeInternal)(false);
    const { update, refs, placement: flPlacement, middlewareData: { arrow: { x: arrowX, y: arrowY } = {} }, floatingStyles, } = (0, react_1.useFloating)({
        strategy: chosenStrategy,
        placement,
        open,
        middleware: [
            (0, react_1.offset)(offset !== null && offset !== void 0 ? offset : (themeContext ? 8 : arrow ? 16 : 4)),
            chosenFlip &&
                (0, react_1.flip)({ padding: 5, fallbackPlacements: ["bottom", "top"] }),
            (0, react_1.shift)({ padding: 12 }),
            (0, react_1.arrow)({ element: arrowRef, padding: 8 }),
        ],
    });
    (0, hooks_1.useClientLayoutEffect)(() => {
        refs.setReference(anchorEl);
    }, [anchorEl]);
    const floatingRef = (0, useMergeRefs_1.useMergeRefs)(refs.setFloating, ref);
    (0, hooks_1.useClientLayoutEffect)(() => {
        if (!refs.reference.current || !refs.floating.current || !open)
            return;
        const cleanup = (0, react_1.autoUpdate)(refs.reference.current, refs.floating.current, update);
        return () => cleanup();
    }, [refs.floating, refs.reference, update, open, anchorEl]);
    const staticSide = {
        top: "bottom",
        right: "left",
        bottom: "top",
        left: "right",
    }[flPlacement.split("-")[0]];
    return (react_2.default.createElement(DismissableLayer_1.DismissableLayer, { asChild: true, safeZone: {
            anchor: anchorEl,
            dismissable: refs.floating.current,
        }, onDismiss: () => open && (onClose === null || onClose === void 0 ? void 0 : onClose()), enabled: open },
        react_2.default.createElement("div", Object.assign({ ref: floatingRef }, rest, { className: cn("navds-popover", className, {
                "navds-popover--hidden": !open || !anchorEl,
            }), style: Object.assign(Object.assign({}, rest.style), floatingStyles), "data-placement": flPlacement, "aria-hidden": !open || !anchorEl }),
            children,
            arrow && !themeContext && (react_2.default.createElement("div", { ref: (node) => {
                    arrowRef.current = node;
                }, style: Object.assign(Object.assign(Object.assign({}, (arrowX != null ? { left: arrowX } : {})), (arrowY != null ? { top: arrowY } : {})), (staticSide ? { [staticSide]: "-0.5rem" } : {})), className: cn("navds-popover__arrow") })))));
});
exports.Popover.Content = PopoverContent_1.default;
exports.default = exports.Popover;
