// @navikt/ds-react/cjs/list/List.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.List = void 0;
const react_1 = __importStar(require("react"));
const Theme_1 = require("../theme/Theme");
const typography_1 = require("../typography");
const List_Item_1 = require("./List.Item");
const List_context_1 = require("./List.context");
const headingSizeMap = {
    small: "xsmall",
    medium: "small",
    large: "medium",
};
/**
 * A list component
 *
 * @see [📝 Documentation](https://aksel.nav.no/komponenter/core/list)
 * @see 🏷️ {@link ListProps | Props}
 *
 * @example
 * ```jsx
 * <List>
 *   <List.Item>Coffee</List.Item>
 *   <List.Item>Tea</List.Item>
 *   <List.Item>Milk</List.Item>
 * </List>
 * ```
 */
exports.List = (0, react_1.forwardRef)((_a, ref) => {
    var { children, className, as: ListTag = "ul", title, description, headingTag, size, "aria-label": _ariaLabel, "aria-labelledby": _ariaLabelledBy } = _a, rest = __rest(_a, ["children", "className", "as", "title", "description", "headingTag", "size", "aria-label", "aria-labelledby"]);
    const { size: contextSize } = (0, react_1.useContext)(List_context_1.ListContext);
    const { cn } = (0, Theme_1.useRenameCSS)();
    const themeContext = (0, Theme_1.useThemeInternal)(false);
    const listSize = size !== null && size !== void 0 ? size : contextSize;
    if (themeContext) {
        if (process.env.NODE_ENV !== "production" &&
            (title || description || headingTag)) {
            console.warn("List: title, description and headingTag are deprecated and will not work with updated theme for Aksel.");
        }
        return (react_1.default.createElement(List_context_1.ListContext.Provider, { value: {
                listType: ListTag,
                size: listSize,
            } },
            react_1.default.createElement(typography_1.BodyLong, Object.assign({ as: "div" }, rest, { size: listSize, ref: ref, className: cn("navds-list", `navds-list--${listSize}`, className) }),
                react_1.default.createElement(ListTag, { role: "list", "aria-label": _ariaLabel, "aria-labelledby": _ariaLabelledBy }, children))));
    }
    return (react_1.default.createElement(List_context_1.ListContext.Provider, { value: {
            listType: ListTag,
            size: listSize,
        } },
        react_1.default.createElement(typography_1.BodyLong, Object.assign({ as: "div" }, rest, { size: listSize, ref: ref, className: cn("navds-list", `navds-list--${listSize}`, className) }),
            title && (react_1.default.createElement(typography_1.Heading, { size: headingSizeMap[listSize], as: headingTag !== null && headingTag !== void 0 ? headingTag : "h3" }, title)),
            description && react_1.default.createElement(typography_1.BodyShort, { size: listSize }, description),
            react_1.default.createElement(ListTag, { role: "list", "aria-label": _ariaLabel, "aria-labelledby": _ariaLabelledBy }, children))));
});
exports.List.Item = List_Item_1.ListItem;
exports.default = exports.List;
