// @navikt/ds-react/cjs/theme/Theme.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useThemeInternal = exports.useRenameCSS = exports.Theme = exports.compositeClassFunction = void 0;
const clsx_1 = __importDefault(require("clsx"));
const react_1 = __importStar(require("react"));
const Slot_1 = require("../slot/Slot");
const create_context_1 = require("../util/create-context");
const [RenameCSSProvider, useRenameCSS] = (0, create_context_1.createContext)({
    hookName: "useRenameCSS",
    name: "RenameCSS",
    providerName: "RenameCSSProvider",
    defaultValue: { cn: clsx_1.default },
});
exports.useRenameCSS = useRenameCSS;
const compositeClassFunction = (...inputs) => {
    const classes = (0, clsx_1.default)(inputs)
        /* Replaces only if start of string  "navds- navds-"*/
        .replace(/^navds-/g, "aksel-")
        /* Replaces all " navds-" hits */
        .replace(/\snavds-/g, " aksel-");
    return classes.trim();
};
exports.compositeClassFunction = compositeClassFunction;
const RenameCSS = ({ children }) => {
    return (react_1.default.createElement(RenameCSSProvider, { cn: exports.compositeClassFunction }, children));
};
const [ThemeProvider, useThemeInternal] = (0, create_context_1.createContext)({
    hookName: "useTheme",
    name: "ThemeProvider",
    providerName: "ThemeProvider",
});
exports.useThemeInternal = useThemeInternal;
const Theme = (0, react_1.forwardRef)((props, ref) => {
    const context = useThemeInternal(false);
    const { children, className, asChild = false, theme = context === null || context === void 0 ? void 0 : context.theme, hasBackground: hasBackgroundProp = true, } = props;
    const isRoot = context === undefined;
    const hasBackground = hasBackgroundProp !== null && hasBackgroundProp !== void 0 ? hasBackgroundProp : (isRoot && props.theme !== undefined);
    const SlotElement = asChild ? Slot_1.Slot : "div";
    return (react_1.default.createElement(ThemeProvider, { theme: theme },
        react_1.default.createElement(RenameCSS, null,
            react_1.default.createElement(SlotElement, { ref: ref, className: (0, clsx_1.default)("aksel-theme", className, theme), "data-background": hasBackground }, children))));
});
exports.Theme = Theme;
