// @navikt/ds-react/cjs/provider/Provider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Provider = exports.useProvider = exports.ProviderContext = void 0;
const react_1 = __importStar(require("react"));
const nb_1 = __importDefault(require("../util/i18n/locales/nb"));
exports.ProviderContext = (0, react_1.createContext)({
    locale: nb_1.default,
});
const useProvider = () => (0, react_1.useContext)(exports.ProviderContext);
exports.useProvider = useProvider;
/**
 * Provides added context to the component tree.
 *
 * @see [📝 Documentation](https://aksel.nav.no/komponenter/core/provider)
 * @see 🏷️ {@link ProviderProps}
 *
 * @example
 * ```jsx
 * <Provider rootElement={rootElement}>
 *   {app}
 * </Provider>
 * ```
 */
const Provider = ({ children, rootElement, locale, translations, }) => {
    const parentContext = (0, exports.useProvider)();
    return (react_1.default.createElement(exports.ProviderContext.Provider, { value: {
            rootElement: rootElement || parentContext.rootElement,
            locale: locale || parentContext.locale || nb_1.default,
            translations: translations || parentContext.translations,
        } }, children));
};
exports.Provider = Provider;
exports.default = exports.Provider;
